lib/pbench/server/api/resources/upload_api.py

```diff
@@ -22,29 +27,174 @@
             )
             abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="INTERNAL ERROR")
-        response.status_code = HTTPStatus.OK
-        return response
+        if status.startswith("MESSAGE==="):
+            msg = status[len("MESSAGE===") :]
+            abort(HTTPStatus.SERVICE_UNAVAILABLE, message=msg)
+        return HTTPStatus.OK
+
+
+class CleanupTime(Exception):
+    """
+    Used to support handling errors during PUT without constantly testing the
+    current status and additional indentation. This will be raised to an outer
+    try block when an error occurs.
+    """
+
+    def __init__(self, status: int, message: str):
+        self.status = status
+        self.message = message
+
+
+class Step(Enum):
+    """
+    Define the persistent changes caused during the PUT upload, which need to
+    be undone if the upload ultimately fails.
+    """
+
+    DATASET = 1  # Dataset was created
+    UPLOAD = 2  # Uploaded tar file was created
+    MD5 = 3  # MD5 file was created
+
+
+class CleanupAction:
+    """
+    Define a single cleanup action necessary to reverse persistent steps in the
+    upload procedure, based on the Step ENUM associated with the action.
+    """
+
+    def __init__(self, step: Step, logger: Logger, parameter: Any):
+        """
+        Define a cleanup action
+
+        Args:
+            step: The persistent step taken
+            logger: The active Pbench Logger object
+            parameter: The object (type varies depending on step) on which the
+                cleanup action is performed.
+        """
+        self.step = step
+        self.parameter = parameter
+        self.logger = logger
+
+    def cleanup(self):
+        """
+        Perform a cleanup action depending on the associated Step value.
+
+        This handles errors and reports them, but doesn't propagate failure to
+        ensure that cleanup continues as best we can.
+        """
+        if self.step is Step.DATASET:
+            try:
+                self.parameter.delete()
+            except Exception as e:
+                self.logger.error("Unable to remove dataset {}: {}", self.parameter, e)
+        elif self.step is Step.MD5:
+            try:
+                self.parameter.unlink()
+            except FileNotFoundError:
+                pass  # Note: Python 3.8 avoids this with `missing_ok=True`
+            except Exception as e:
+                self.logger.error(
+                    "Unable to remove temporary MD5 file {}: {}", self.parameter, e
+                )
+        elif self.step is Step.UPLOAD:
+            try:
+                self.parameter.unlink()
+            except FileNotFoundError:
+                pass  # Note: Python 3.8 avoids this with `missing_ok=True`
+            except Exception as e:
+                self.logger.error(
+                    "Unable to remove temporary TAR file {}: {}", self.parameter, e
+                )
```

Comment:
I considered that, but they're fairly small and this seemed more direct and easier to follow. (Sigh.)

Reply:
The fact that they are fairly small means that they aren't hard to construct in the caller's context.  (Unfortunately, the `try` blocks preclude using lambdas, but maybe we could have the cleanup mechanism catch the exceptions instead of doing so in the lambdas themselves.)

Re "easy to follow", the problem is that there's an "action at a distance" effect, where the cleanup code is in some far-off function instead of attached to the code which created the context that needs to be cleaned up.  So, yeah, its easy to follow...if you have multiple windows...but it otherwise lacks context, and it's sort of pre-canned.